### PR TITLE
Moving msbuild over to Arcade

### PR DESCRIPTION
**Pull request by livarcocc**

### CODE REVIEW

**Comment by livarcocc** [comment]
@rainersigwald @dsplaisted this should be ready for review. Would you folks like to start looking through it?

File: src/Build.OM.UnitTests/Construction/ConstructionEditing_Tests.cs [hunk 1]
```diff
@@ -40,13 +40,13 @@ public void AddTargetConvenience()
   <Target Name=""t"" />
 </Project>");
 
-            Assert.Equal(true, project.HasUnsavedChanges);
+            Assert.True(project.HasUnsavedChanges);
```
Comment (rainersigwald) [inline src/Build.OM.UnitTests/Construction/ConstructionEditing_Tests.cs]: Was this whole commit suggested fixes, or did you have to do anything by hand?

File: src/Build.UnitTests/BackEnd/AssemblyTaskFactory_Tests.cs [hunk 2]
```diff
@@ -659,7 +659,7 @@ public void VerifySameFactoryCanGenerateDifferentTaskInstances()
 #endif
                     false);
                 Assert.NotNull(createdTask);
-                Assert.IsType<TaskHostTask>(createdTask);
+                Assert.IsNotType<TaskHostTask>(createdTask);
```
Comment (rainersigwald) [inline src/Build.UnitTests/BackEnd/AssemblyTaskFactory_Tests.cs]: This one worries me -- it seems to have been passing with the previous assertion. What was the actual type?

Comment (rainersigwald) [reply src/Build.UnitTests/BackEnd/AssemblyTaskFactory_Tests.cs]: Never mind, fixed in a later commit.

File: build.cmd [hunk 3]
```diff
@@ -1,3 +1,3 @@
 @echo off
-powershell -NoLogo -NoProfile -ExecutionPolicy ByPass -Command "& """%~dp0eng\common\build.ps1""" -build -restore -p:Projects="""%~dp0MSBuild.sln""" %*"
+powershell -NoLogo -NoProfile -ExecutionPolicy ByPass -Command "& """%~dp0eng\common\build.ps1""" -build -restore /p:Projects="""%~dp0MSBuild.sln""" %*"
```
Comment (rainersigwald) [inline build.cmd]: Was this really necessary? MSBuild should have supported both; is it a powershell problem?

File: src/Package/MSBuild.Engine.Corext/MSBuild.Engine.Corext.csproj [hunk 4]
```diff
@@ -1,4 +1,13 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
+  <!-- Don't try to build this package if building for Mono, using .NET Core, or building on non-Windows -->
+  <!-- Don't try to build this package if building for Mono, using .NET Core, or building on non-Windows -->
```
Comment (rainersigwald) [inline src/Package/MSBuild.Engine.Corext/MSBuild.Engine.Corext.csproj]: nit: doubled

File: src/Build.OM.UnitTests/Construction/ConstructionEditing_Tests.cs (hunk 1 above)
Comment (livarcocc) [reply src/Build.OM.UnitTests/Construction/ConstructionEditing_Tests.cs]: All suggested fixes.

File: src/Build.UnitTests/BackEnd/AssemblyTaskFactory_Tests.cs (hunk 2 above)
Comment (livarcocc) [reply src/Build.UnitTests/BackEnd/AssemblyTaskFactory_Tests.cs]: Right.

File: build.cmd (hunk 3 above)
Comment (livarcocc) [reply build.cmd]: Powershell complained about it. Not a msbuild problem.

File: eng/common/cross/android/arm/toolchain.cmake [hunk 5]
```diff
@@ -0,0 +1,41 @@
+set(CROSS_NDK_TOOLCHAIN $ENV{ROOTFS_DIR}/../)
```
[the remark this answers is not in the document]
Comment (rainersigwald) [reply eng/common/cross/android/arm/toolchain.cmake]: Yeah, it's fine. I'll throw a `Makefile` in the root too just to keep everybody on their toes :)

**Comment by rainersigwald** [comment]
Another concern: This PR (https://dnceng.visualstudio.com/public/_build/results?buildId=55637&view=ms.vss-test-web.test-result-details) ran 35,064 tests. Another recent PR to master (https://dnceng.visualstudio.com/public/_build/results?buildId=55416&view=ms.vss-test-web.test-result-details) ran 47,247. What's causing the discrepancy?

File: build/build.ps1 [hunk 6]
```diff
@@ -300,7 +300,7 @@ function Build {
       # - Don't pack
       # - Do run tests (if not skipped)
       # - Don't try to create a bootstrap deployment
-      CallMSBuild `"$RepoToolsetBuildProj`" @msbuildArgs /nr:false /p:Restore=$restore /p:Build=$build /p:Rebuild=$rebuild /p:Test=$runTests /p:Sign=false /p:Pack=false /p:CreateBootstrap=false @properties
+      CallMSBuild `"$RepoToolsetBuildProj`" @msbuildArgs /nr:false /p:Restore=$restore /p:Build=$build /p:Rebuild=$rebuild /p:Test=$runTests /p:Sign=false /p:Pack=false /p:CreateBootstrap=false /p:UsingToolMicrosoftNetCompilers=false @properties
```
[the remark this answers is not in the document]
Comment (livarcocc) [reply build/build.ps1]: Deleted.

File: build/cibuild.cmd [hunk 7]
```diff
@@ -1,3 +1,3 @@
 @echo off
-powershell -NoLogo -NoProfile -ExecutionPolicy ByPass "& '%~dp0build.ps1'" -build -bootstrap -pack -sign -ci -prepareMachine %*
+powershell -NoLogo -NoProfile -ExecutionPolicy ByPass "& '%~dp0build.ps1'" -build -bootstrap -pack -sign -ci -prepareMachine -publish %*
```
[the remark this answers is not in the document]
Comment (livarcocc) [reply build/cibuild.cmd]: I just avoided moving stuff over from build to eng to reduce to amount of changes. I will add a subsequent commit deleting all the stuff from build that we don't need anymore. I will also take a stab and moving the files over to eng.

**Comment by livarcocc** [comment]
Regarding your last comment about the number of tests, I would imagine the difference there is in the fact that I turned off the HostType CI leg., for now.

> We need a (new?) official build definition with a run and a test insertion with DDRITs and RPS.

Yes. My plan is to work on that next after this PR gets merged.

> Are we losing the bootstrap build? It look like the CI tests were run based on the first-stage build which is fine, but I don't want to lose the second-stage build; it's caught real problems. I'd be open to adding it as a new CI phase, which might reduce turnaround time and reduce some mysterious errors.

Can you help me understand what is special about these tests? Do we need anything special to kick them off? I would expect those tests to simply run if they are part of the solution.

> It seems like there's a bunch of semi-irrelevant junk in the eng/ directory; should we just deal with it for uniformity or can we scope down?

Right, as I commented above, this is how arcade works. The contents of eng/common come straight from arcade and we can't pick and choose.

> I wish the test-analyzers stuff could be decoupled, but oh well.

I guess I could have done that by disabling the analyzers, but I prefered to fix them.

File: src/Package/MSBuild.Engine.Corext/MSBuild.Engine.Corext.csproj (hunk 4 above)
Comment (livarcocc) [reply src/Package/MSBuild.Engine.Corext/MSBuild.Engine.Corext.csproj]: Fixed.

File: Directory.Build.props [hunk 8]
```diff
@@ -55,7 +55,7 @@
         you don't get those defaults.
     -->
 
-    <NoWarn>$(NoWarn);NU1603;NU5105;1701;1702</NoWarn>
+    <NoWarn>$(NoWarn);NU1603;NU5105;NU5125;1701;1702</NoWarn>
```
[the remark this answers is not in the document]
Comment (livarcocc) [reply Directory.Build.props]: Done.

**Comment by dsplaisted** [comment]
Also, as far as the contents of the `eng/common` folder go, once we have dependency flow from arcade to MSBuild hooked up, these files will get sync'd automatically with the arcade versions whenever the arcade version in the repo is updated (typically daily).

File: .vsts-dotnet-ci.yml [hunk 9]
```diff
@@ -31,166 +26,269 @@ phases:
...
     inputs:
-      PathtoPublish: 'artifacts/2/Debug/log/TestResults'
+      PathtoPublish: 'artifacts/TestResults'
       ArtifactName: 'FullOnWindows test logs'
     condition: succeededOrFailed()
 
-- phase: FullReleaseOnWindows
-  displayName: "Build and test Release on Windows using full MSBuild"
+- phase: BootstrapMSBuildOnFullFrameworkWindows
+  displayName: "Build and test on Windows using bootstrapped full MSBuild"
```
[the remark this answers is not in the document]
Comment (livarcocc) [reply .vsts-dotnet-ci.yml]: We only test on the bootstrapped one.

File: .vsts-dotnet.yml [hunk 10]
```diff
@@ -1,142 +1,92 @@
...
+  - script: eng/common/cibuild.cmd
+              -configuration $(BuildConfiguration)
+              $(_PublishArgs)
+              $(_SignArgs)
+              /p:OfficialBuildId=$(BUILD.BUILDNUMBER)
+    displayName: Build
+    condition: succeeded()
+
+  - task: PublishTestResults@2
+    displayName: Publish .NET Framework Test Results
```
[the remark this answers is not in the document]
Comment (livarcocc) [reply .vsts-dotnet.yml]: This is not the official YAML. This is the CI one.

File: eng/Build.props [hunk 11]
```diff
@@ -2,11 +2,13 @@
...
-  <Import Project="Sdk.props" Sdk="RoslynTools.RepoToolset"/>
-
-  <Target Name="Build"/>
+  <ItemGroup>
+    <ProjectToBuild Remove="@(ProjectToBuild)" />
+    <ProjectToBuild Include="$(RepoRoot)MSBuild.sln" />
+  </ItemGroup>
+
+  <PropertyGroup>
+    <!-- <Projects>$(RepoRoot)MSBuild.sln</Projects> -->
```
[the remark this answers is not in the document]
Comment (livarcocc) [reply eng/Build.props]: No. Left over. I will remove in a clean-up PR following this one up.

File: eng/Build.props [hunk 12]
```diff
@@ -2,11 +2,13 @@
...
 
-  <Import Project="ProducesNoOutput.Settings.props" />
-
-  <!-- The nuget resolver will download the sdk -->
-  <Import Project="Sdk.props" Sdk="RoslynTools.RepoToolset"/>
-
-  <Target Name="Build"/>
+  <ItemGroup>
+    <ProjectToBuild Remove="@(ProjectToBuild)" />
+    <ProjectToBuild Include="$(RepoRoot)MSBuild.sln" />
```
[the remark this answers is not in the document]
Comment (livarcocc) [reply eng/Build.props]: Ok to do in a follow up PR?

[the remark this answers is not in the document]
Comment (rainersigwald) [reply eng/Build.props]: Yeah, that works for me.

File: .vsts-dotnet.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (livarcocc) [reply .vsts-dotnet.yml]: Nvm, I see now this is on .vsts-dotnet.yml. Yes, we can change this as well to skip tests

**Comment by livarcocc** [comment]
@rainersigwald I checked that the vsman is now being generated.

**Pull request merged**